harvester/harvester
# Support storage tiering

Signed-off-by: futuretea <Hang.Yu@suse.com>

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

HEP: https://github.com/harvester/harvester/pull/2335

**Related Issue:**
https://github.com/harvester/harvester/issues/2147

**Test plan:**


Build images:
```
export REPO=<your registry repo name>
export PUSH=true
make
```
Or use my prebuild image: `futuretea/harvester:st-head` and `futuretea/harvester-webhook:st-head`

Need a Harvester cluster in v1.0.2 version and pre-create some images in the cluster for test backward compatibility.

Upgrade CRDs:
```
cd deploy/charts/harvester-crd
helm upgrade harvester-crd ./ -n harvester-system
```

Upgrade Harvester components:
```
kubectl -n harvester-system set image deployment/harvester apiserver=futuretea/harvester:st-head
kubectl -n harvester-system set image deployment/harvester-webhook harvester-webhook=futuretea/harvester-webhook:st-head

```

in one single node cluster, need a force harvester-webhook restart 
```
kubectl -n harvester-system scale deploy harvester-webhook --replicas=0
kubectl -n harvester-system scale deploy harvester-webhook --replicas=1
```

Add node/disk tags  using the Longhorn UI https://docs.harvesterhci.io/v1.0/troubleshooting/harvester/#access-embedded-longhorn

1. Node -> Select one node -> Edit Node and Disks
2. Click +New Node Tag or +New Disk Tag to add new tags.

Test cases:
1. Old version images and storageClasses and backingimages should not be changed.

2. Edit any old version image, change Description from UI

- The storageClass and backingimage of the image should not be changed.

3. Edit any old version image, add some parameters to `spec.storageClassParameters`.

- this request should be refused by harvester-webhook

4. Create a new image, do not specify `metadata.Annotations["harvesterhci.io/storageClassName"]`

- The field `spec.storageClassParameters` in the image will be mutated and the value is the same as the paramters value of the default storageClass `longhorn` 
- The filed `parameters` in the backingimage storageClass should be the same as `spec.storageClassParameters` in image（key-value "backingimage: xxx" only in the backingimage storageClass )

5. Create a new image, specify `metadata.Annotations["harvesterhci.io/storageClassName"]`
- The field `spec.storageClassParameters` in the image will be mutated and the value is the same as the paramters value of the user specified storageClass 
- The filed `parameters` in the backingimage storageClass should be the same as `spec.storageClassParameters` in image （key-value "backingimage: xxx" only in the backingimage storageClass )

6. create a volume by choose source=image, export the image volume to a new image via `View API` and specify `storageClassName` 

-  the new exported image's `spec.storageClassParameters` should be the same as the paramters value of the storageClass which we specify during the volume export action

7. restore a volume snapshot to a new volume via `View API` and specify `storageClassName` 

-  the new restore volumes `spec.storageClassName` should be the same as the value which we specify during the volume snapshot restore action
